### ISSUE TITLE
Game state machine rewrite

### ISSUE THREAD

**Issue by DarkMatters** [VarynInc/varyn#245]
Current `GameModel` object offers a very dumb finite state machine with 1 method `changeGameState(newState)`. We should improve this to something more useful and reusable.

#### Constructor

Construct a new state machine and set it to an initial state. If `initialState` is not provided then the state machine starts at the first state in the state table.

```
var gameState = new FiniteStateMachine(stateTable, initialState);
```

Create a new state machine from a state transition table. The table is an object that defines the states and what state transition are allowed.

#### State table

The state table defines all available states and what states each state may transition to.

```
{
   init: ["splash", "load"],
   splash: ["load", "play"],
   load: ["play"],
   play: ["menu", "interstitial", "gameover"],
   menu: ["-", "options", "awards", "leaderboard", "credits", "moregames", "share"]
}
```

Note the special state identifier "-". This means the state is allowed to transition back to the prior state that transitioned to this state.

If a state does not appear in the table definition (such as the case for "options" in the above example) then those states will default to "-", which means they are allowed to transition back to their prior state but not any other state.

#### Transition to a new state

The state machine transitions to a new state by calling `transition(newState)`. A transition is a promise to transition to that state and the promise is resolved after the state transition is complete. State transitions may not happen right away. or happen at all, since pre and post events can interrupt the transition and the transition itself takes time and may include asynchronous sub-tasks.

```
   gameState.transition("options").then(resolve, reject);
```

You can also return the state machine back to its prior state if that transition is allowed by the state table.

```
   gameState.transition("-").then(resolve, reject);
```

#### Event listeners

Register event listener functions for pre and post state transitions.

```
   gameState.on(event, state, eventHandler);
```

`event` can be either "pre" or "post". "pre" event is fired before the state transition occurs. "post" is fired after the state transition is complete.

`state` is the state in the state table to listen for.

`eventHandler` is the function to call. The function takes a single parameter, `event` that is an object with the event information. `event.state` is the state that is the subject of the transition. `event.target` is the `FiniteStateMachine` object.

For example:

```
   gameState.on("pre", "interstitial", onPrerollAd(transitionEvent) { 
      // do things before going to the interstitial state, such as load the next ad
   });
```

#### Helper methods

Return the current state. State is read only, this value can only be set by calling `transition`.

```
   var state = gameState.state();
```

Determine if the FSM can transition to the proposed state from the current state it is in.

```
   if (gameState.canTransition(proposedState)) {
```

Determine if the FSM is in the midst of a state transition. Since transitions are asynchronous, you may want to know if some prior transition has not yet completed.

```
   if ( ! gameState.isTransitioning()) {
```

Return a list (array) of allowed transition states from the current state.

```
   const possibleNextStates = gameState.listTransitions();
```

**Comment by varyndev** [comment]
yes, this is important, but it is not critical to the PWA MVP for #108. Keeping high priority but removing it from the milestone.
